### ISSUE TITLE
monomorph issue with composite enum type and projection

### ISSUE THREAD

**Issue by JonathanStarup** [flix/flix#2568]
This program
```
pub enum E {}
pub def f(x: List[(String, E)]): Unit = let _x = project x into X; ()
```
String can be changed to Int etc. but a type variable doesn't result in the error.
The `List[(String/Int, E)]` is important `List[E]` is fine, so is `List[(E)]`
results in this error
```
[/] [Monomorph] List.reverseHelper$105469
Exception in thread "main" ca.uwaterloo.flix.util.InternalCompilerException: No default or matching definition found for 'Order.compare'.
	at ca.uwaterloo.flix.language.phase.Monomorph$.specializeSigSym$1(Monomorph.scala:608)
	at ca.uwaterloo.flix.language.phase.Monomorph$.visitExp$1(Monomorph.scala:185)
	at ca.uwaterloo.flix.language.phase.Monomorph$.visitExp$1(Monomorph.scala:246)
	at ca.uwaterloo.flix.language.phase.Monomorph$.visitExp$1(Monomorph.scala:242)
	at ca.uwaterloo.flix.language.phase.Monomorph$.visitExp$1(Monomorph.scala:242)
	at ca.uwaterloo.flix.language.phase.Monomorph$.visitExp$1(Monomorph.scala:266)
	at ca.uwaterloo.flix.language.phase.Monomorph$.visitExp$1(Monomorph.scala:266)
	at ca.uwaterloo.flix.language.phase.Monomorph$.specialize$1(Monomorph.scala:562)
	at ca.uwaterloo.flix.language.phase.Monomorph$.$anonfun$run$1(Monomorph.scala:765)
	at ca.uwaterloo.flix.api.Flix.phase(Flix.scala:421)
	at ca.uwaterloo.flix.language.phase.Monomorph$.run(Monomorph.scala:128)
	at ca.uwaterloo.flix.language.phase.Monomorph$.run(Monomorph.scala:58)
	at ca.uwaterloo.flix.language.phase.Phase$$anon$1.$anonfun$run$1(Phase.scala:42)
	at ca.uwaterloo.flix.util.Validation.flatMap(Validation.scala:50)
	at ca.uwaterloo.flix.util.Validation.flatMap$(Validation.scala:49)
	at ca.uwaterloo.flix.util.Validation$Success.flatMap(Validation.scala:74)
	at ca.uwaterloo.flix.language.phase.Phase$$anon$1.run(Phase.scala:41)
	at ca.uwaterloo.flix.language.phase.Phase$$anon$1.run(Phase.scala:41)
	at ca.uwaterloo.flix.language.phase.Phase$$anon$1.run(Phase.scala:41)
	at ca.uwaterloo.flix.language.phase.Phase$$anon$1.run(Phase.scala:41)
	at ca.uwaterloo.flix.language.phase.Phase$$anon$1.run(Phase.scala:41)
	at ca.uwaterloo.flix.language.phase.Phase$$anon$1.run(Phase.scala:41)
	at ca.uwaterloo.flix.language.phase.Phase$$anon$1.run(Phase.scala:41)
	at ca.uwaterloo.flix.language.phase.Phase$$anon$1.run(Phase.scala:41)
	at ca.uwaterloo.flix.language.phase.Phase$$anon$1.run(Phase.scala:41)
	at ca.uwaterloo.flix.language.phase.Phase$$anon$1.run(Phase.scala:41)
	at ca.uwaterloo.flix.language.phase.Phase$$anon$1.run(Phase.scala:41)
	at ca.uwaterloo.flix.language.phase.Phase$$anon$1.run(Phase.scala:41)
	at ca.uwaterloo.flix.language.phase.Phase$$anon$1.run(Phase.scala:41)
	at ca.uwaterloo.flix.api.Flix.codeGen(Flix.scala:388)
	at ca.uwaterloo.flix.api.Flix.$anonfun$compile$1(Flix.scala:405)
	at ca.uwaterloo.flix.util.Validation.flatMap(Validation.scala:50)
	at ca.uwaterloo.flix.util.Validation.flatMap$(Validation.scala:49)
	at ca.uwaterloo.flix.util.Validation$Success.flatMap(Validation.scala:74)
	at ca.uwaterloo.flix.api.Flix.compile(Flix.scala:404)
	at ca.uwaterloo.flix.Main$.$anonfun$main$5(Main.scala:186)
	at ca.uwaterloo.flix.util.Timer.<init>(Timer.scala:28)
	at ca.uwaterloo.flix.Main$.main(Main.scala:186)
	at ca.uwaterloo.flix.Main.main(Main.scala)

Process finished with exit code 1
```

**Comment by mlutze** [comment]
This is a type class implementation issue.

`Boxable` has type constraint `Order`
`Boxable[(a, b)]` has no type constraints.
`Order[(a, b)]` has constraints `Order[a]` and `Order[b]`

When we check the `Boxable[(a, b)]` instance, we make sure there's an `Order[(a, b)]` instance. 
The issue is that we don't check that the `Boxable[(a, b)]` instance's type constraints entail the `Order[(a, b)]` instance's type constraints.

tl;dr: `Boxable[(a, b)]` needs `Order[a]`, `Order[b]`, `ToString[a]`, `ToString[b]` type constraints and the compiler needs to enforce this.
